keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,795 @@
+A single-item List or Watch request is remapped to the Get verb, so Get requests can have field and label selectors.
+
+### Authorization Attributes changes
+
+The authorization attributes have easy access to the query parameter field and label selectors.
+To avoid confusion, field and label selectors will not be included in authorization attributes for verbs where the field
+selector has no semantic meaning.
+In practice this means that (for now), only Get, List, Watch, and DeleteCollection have field and label selectors.
+Any authorizer that gets an error from `ParseFieldSelector` or `ParseLabelSelector` may attempt to authorize with the 
+`GetFieldSelector` or `GetLabelSelector` or attempt to authorize without field or label selectors since that will authorize
+using a wider permission (field and label selectors can only reduce access).
+
+```go
+type Attributes interface {
+	// GetFieldSelector returns the query parameter value of the field selector
+	// Remember that only List, Watch, and DeleteCollection (currently) provide 
+	// Remember that field selector formats vary based on the version of the API being used!
+	GetFieldSelector() string
+
+    // GetLabelSelector returns the query parameter value of the label selector
+	GetLabelSelector() string
+
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+	// It can return an error if the field selector cannot be parsed.
+    // Remember that field selector formats vary based on the version of the API being used!
+	ParseFieldSelector() ([]Requirement, error)
+  
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+    // It can return an error if the field selector cannot be parsed.
+    ParseLabelSelector() ([]Requirement, error)
+```
+
+### SubjectAccessReview Changes
+
+SubjectAccessReview is use for two purposes:
+1. Authorization webhook calls from the kube-apiserver to a webhook.
+   This usage likely benefits from a serialization with `[]Requirement`.
+2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)
+   This usage likely benefits from a serialization that matches the query parameter.
+
+Their needs best met with two different serialization (see user stories)
+
+```go
+type SubjectAccessReviewSpec struct {
+	FieldSelector *FieldSelectorAttributes
+	
+	LabelSelector *LabelSelectorAttributes
+}
+
+// FieldSelectorAttributes indicates a field limited access.
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the rawSelector should be honored.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the implementation must describe which (or both) is honored.
+// Remember, the rawSelector may not match the requirements!
```

Comment:
What is the usecase for these two not matching each other?

Reply:
> What is the usecase for these two not matching each other?

No use-case, but recognition that clients can place any data inside a request and not all combinations make sense.  Webhook authorizers are likely to pay the cost to reparse and validate the input since the only impacted party is the requestor.

Reply:
the kube-apiserver can control what it sends and what it accepts, but general webhook authors cannot control what is sent from all who have access to it.  Even if we spec them to match, a naughty client can lie and authors should be expecting to handle such a case.

Reply:
>This eliminates the mismatch edge and ensures webhooks don't do naive incorrect checks against the raw string
> require they match if both are set

We can say in the doc it is required.  We cannot ensure that a naughty client doesn't directly send a mismatched request to a webhook.

Reply:
as long as kube-apiserver always transforms raw strings into parsed requirements (if they parse), and never sends a raw string to a webhook in a SAR, I'm fine with letting naughty clients and clever webhooks those naughty clients talk directly to do whatever they want

Reply:
I would tell webhook authors to only look at requirements and authorize any SAR that doesn't include requirements as an unfiltered request